mds/apis/agency_api/v0_x/compliance.py

```diff
@@ -0,0 +1,119 @@
+from rest_framework import serializers
+from rest_framework import viewsets
+
+from django.db.models import Prefetch
+
+from mds import models
+
+from datetime import datetime
+import pytz
+
+
+class ComplianceSerializer(serializers.ModelSerializer):
+    rules_id = serializers.SerializerMethodField()
+    compliances = serializers.SerializerMethodField()
+
+    class Meta:
+        model = models.Policy
+        fields = ("id", "rules_id", "compliances")
+
+    def get_rules_id(self, policy):
+        return [rule["rule_id"] for rule in policy.rules]
+
+    def get_compliances(self, policy):
+        final_compliance = []
+        for compliance in policy.compliances_pref:
+            for current_compliance in final_compliance:
+                if current_compliance["rule_id"] == compliance.rule:
+                    for match in current_compliance["matches"]:
+                        if match["geography"] == str(compliance.geography):
+                            match["measured"] += 1
+                            break
+                    else:
+                        current_compliance["matches"].append(
+                            {"geography": str(compliance.geography), "measured": 1}
+                        )
+                    current_compliance["vehicles_in_violation"].append(
+                        str(compliance.vehicle.id)
+                    )
+                    current_compliance["total_violations"] = len(
+                        current_compliance["vehicles_in_violation"]
+                    )
+                    break
+            else:
+                final_compliance.append(
+                    {
+                        "rule_id": compliance.rule,
+                        "matches": [
+                            {"geography": str(compliance.geography), "measured": 1}
+                        ],
+                        "vehicles_in_violation": [str(compliance.vehicle.id)],
+                        "total_violations": 1,
+                    }
+                )
+        return final_compliance
+
+
+class ComplianceViewSet(viewsets.ModelViewSet):
+    queryset = models.Policy.objects
+    serializer_class = ComplianceSerializer
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        provider_id = self.request.GET.get("provider_id")
+        end_date = self.request.GET.get("end_date")
+        if end_date:
+            end_date = datetime.fromtimestamp(int(end_date), tz=pytz.UTC)
```

Comment:
.exclude(...date_lt=...) is the same as .filter(...date_gte=...) 

So you could make a dictionnary in your ifs and pass **filter_dict to your filter

```
filter_dict = {"compliances__end_date__gte": end_date, "compliances__vehicle__provider__id": provider_id...}
queryset.filter(**filter_dict)
```

Reply:
nope :)
because end_date may be null
=> @ppoursafa you should maybe add or adapt test to be sure that the case with a object with end_date=None is handled correctly, to prevent regressions in the future

Reply:
None is handled correctly so i let exclude ;)